Python/aot_ceval_jit.c

```diff
@@ -2486,35 +2489,48 @@ static int emit_special_binary_op_refcnt1(Jit* Dst, int inst_idx, int opcode, in
         default:
             return -1;
     }
-    _PyOpcache* co_opcache = get_opcache_entry(Dst->opcache, inst_idx);
-    if (!co_opcache || !co_opcache->optimized) {
-        return -1;
-    }
-    _PyOpcache_TypeRefcnt* cache = &co_opcache->u.t_refcnt;
-    if (!cache->refcnt1_left && !cache->refcnt1_right) {
+    _PyOpcache* opcache = get_opcache_entry(Dst->opcache, inst_idx);
+    if (!opcache || !opcache->optimized) {
         return -1;
     }
+    _PyOpcache_TypeRefcnt* cache = &opcache->u.t_refcnt;
     if (cache->type != &PyFloat_Type) {
         return -1;
     }
-    int use_left = cache->refcnt1_left > cache->refcnt1_right && ref_status[1] == OWNED;
-    // the inplace modified reference must be owned otherwise the refcnt==1 does not mean it's temporary.
-    if (ref_status[use_left ? 1 : 0] != OWNED)
-        return -1;
-    // some simple heuristics: if it looks like the refcnt is only 1 in less than half the cases
-    // don't inline it
-    if ((use_left ? cache->refcnt1_left : cache->refcnt1_right) < co_opcache->optimized/2)
+    // if the left operand is a fast variable and the result is stored in it we can do a special
+    // optimization if we know it's not stored anywhere else.
+    // In the interpreter this cases have refcnt == 2 because the variable is stored in the
+    // fast arg array and on the value stack.
+    // if profiling showed that the right operand can be inplace modifed more often use it instead
+    int load_store_left = load_store_left_idx != -1 && cache->refcnt2_left >= cache->refcnt1_right;
+    if (!cache->refcnt1_left && !cache->refcnt1_right && !load_store_left) {
         return -1;
+    }
+    // should we inplace modify the right operand?
+    int use_right = !load_store_left && ref_status_right == OWNED && cache->refcnt1_right >=  cache->refcnt1_left && cache->refcnt1_right >= opcache->optimized/2;
+    if (!use_right) { // try the left
+        // the inplace modified reference must be owned otherwise the refcnt==1 does not mean it's temporary.
+        if (ref_status_left != OWNED)
```

Comment:
For `a += b`, won't the lhs be a borrowed reference?

Reply:
Good catch! fixing this increased the number of hits quite a lot on some benchmarks